File: Examples/GroupDocs.Annotation.Examples.VisualBasic/GroupDocs.AnnotationDataStorage/Repositories/JsonRepository.vb
Imports System.Collections.Generic
Imports System.Linq
Imports GroupDocs.Annotation.Handler.Input
Imports GroupDocs.Annotation.Handler.Input.DataObjects

Namespace GroupDocs.Data.Json.Repositories
    Public Class JsonRepository(Of TEntity As {Class, IEntity})
        Inherits JsonFile(Of List(Of TEntity))
        Implements IRepository(Of TEntity)
#Region "Fields"
        Private Shared ReadOnly _entityName As String = GetType(TEntity).Name
#End Region

        Public Sub New(filePath As String)
            MyBase.New(filePath)
        End Sub

        Public Sub Commit() Implements IRepository(Of TEntity).Commit

            Serialize()
        End Sub

        Public Sub Refresh(entity As TEntity) Implements IRepository(Of TEntity).Refresh
            Deserialize()
        End Sub

        Public Function Add(entity As TEntity) As Boolean Implements IRepository(Of TEntity).Add
            SyncLock _syncRoot
                Try
                    Dim data = Me.Data

                    entity.Id = GetNextId()
                    data.Add(entity)

                    Commit()
                    Return True
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to add entity: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function Remove(entity As TEntity) As Boolean Implements IRepository(Of TEntity).Remove
            SyncLock _syncRoot
                Try
                    Data.RemoveAll(Function(e) e.Id = entity.Id)
                    Commit()

                    Return True
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to remove entity: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function Update(entity As TEntity) As Boolean Implements IRepository(Of TEntity).Update
            SyncLock _syncRoot
                Try
                    Dim data = Me.Data
                    Dim index = data.FindIndex(Function(x) x.Id = entity.Id)

                    If index >= 0 Then
                        data(index) = entity
                        Commit()

                        Return True
                    End If

                    Return False
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to update entity: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function Add(entities As IEnumerable(Of TEntity)) As Boolean Implements IRepository(Of TEntity).Add
            SyncLock _syncRoot
                Try
                    Data.AddRange(entities)
                    Commit()

                    Return True
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to add entities: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function Remove(entities As IEnumerable(Of TEntity)) As Boolean Implements IRepository(Of TEntity).Remove
            SyncLock _syncRoot
                Try
                    Dim data = Me.Data

                    For Each e As TEntity In entities
                        data.Remove(e)
                    Next

                    Commit()
                    Return True
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to remove entities: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function Update(entities As IEnumerable(Of TEntity)) As Boolean Implements IRepository(Of TEntity).Update
            SyncLock _syncRoot
                Try
                    Dim data = Me.Data

                    For Each e As TEntity In entities
                        Dim index = data.FindIndex(Function(x) x.Id = e.Id)
                        If index >= 0 Then
                            data(index) = e
                        End If
                    Next

                    Commit()
                    Return True
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to update entities: {0}", _entityName), e)
                End Try
            End SyncLock
        End Function

        Public Function [Get](id As Decimal) As TEntity
            SyncLock _syncRoot
                Try
                    Return Data.Find(Function(e) e.Id = id)
                Catch e As Exception
                    Throw New DataJsonException([String].Format("Unable to get entity: ID = {0}", id), e)
                End Try
            End SyncLock
        End Function

        Public Function [Get](id As Long) As TEntity Implements IRepository(Of TEntity).Get
            Return [Get](CDec(id))
        End Function

        Protected Overridable Function GetNextId(Optional increment As Integer = 1) As Long
            Dim data__1 = Data
            Dim lastId = (If(data__1.Any(), data__1.Max(Function(e) e.Id), 0L))
            Return (lastId + increment)
        End Function

#Region "Private members"
#End Region
    End Class
End Namespace

File: Showcases/GroupDocs-Annotation-Webforms-Front-End/GroupDocs.Data.Json/Repositories/JsonRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using GroupDocs.Annotation.Handler.Input;
using GroupDocs.Annotation.Handler.Input.DataObjects;

namespace GroupDocs.Data.Json.Repositories
{
    public class JsonRepository<TEntity> : JsonFile<List<TEntity>>, IRepository<TEntity>
        where TEntity : class, IEntity
    {
        #region Fields
        private static readonly string _entityName = typeof(TEntity).Name;
        #endregion Fields

        public JsonRepository(string filePath)
            : base(filePath)
        {
        }

        public void Commit()
        {
            Serialize();
        }

        public void Refresh(TEntity entity)
        {
            Deserialize();
        }

        public bool Add(TEntity entity)
        {
            lock (_syncRoot)
            {
                try
                {
                    var data = this.Data;

                    entity.Id = GetNextId();
                    data.Add(entity);

                    Commit();
                    return true;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                        String.Format("Unable to add entity: {0}", _entityName), e);
                }
            }
        }

        public bool Remove(TEntity entity)
        {
            lock (_syncRoot)
            {
                try
                {
                    Data.RemoveAll(e => e.Id == entity.Id);
                    Commit();

                    return true;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                       String.Format("Unable to remove entity: {0}", _entityName), e);
                }
            }
        }

        public bool Update(TEntity entity)
        {
            lock (_syncRoot)
            {
                try
                {
                    var data = this.Data;
                    var index = data.FindIndex(x => x.Id == entity.Id);

                    if (index >= 0)
                    {
                        data[index] = entity;
                        Commit();

                        return true;
                    }

                    return false;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                        String.Format("Unable to update entity: {0}", _entityName), e);
                }
            }
        }

        public bool Add(IEnumerable<TEntity> entities)
        {
            lock (_syncRoot)
            {
                try
                {
                    Data.AddRange(entities);
                    Commit();

                    return true;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                       String.Format("Unable to add entities: {0}", _entityName), e);
                }
            }
        }

        public bool Remove(IEnumerable<TEntity> entities)
        {
            lock (_syncRoot)
            {
                try
                {
                    var data = this.Data;

                    foreach (var e in entities)
                    {
                        data.Remove(e);
                    }

                    Commit();
                    return true;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                        String.Format("Unable to remove entities: {0}", _entityName), e);
                }
            }
        }

        public bool Update(IEnumerable<TEntity> entities)
        {
            lock (_syncRoot)
            {
                try
                {
                    var data = this.Data;

                    foreach (var e in entities)
                    {
                        var index = data.FindIndex(x => x.Id == e.Id);
                        if (index >= 0)
                        {
                            data[index] = e;
                        }
                    }

                    Commit();
                    return true;
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                        String.Format("Unable to update entities: {0}", _entityName), e);
                }
            }
        }

        public TEntity Get(decimal id)
        {
            lock (_syncRoot)
            {
                try
                {
                    return Data.Find(e => e.Id == id);
                }
                catch (Exception e)
                {
                    throw new DataJsonException(
                        String.Format("Unable to get entity: ID = {0}", id), e);
                }
            }
        }

        public TEntity Get(long id)
        {
            return Get((decimal) id);
        }

        protected virtual long GetNextId(int increment = 1)
        {
            var data = Data;
            var lastId = (data.Any() ? data.Max(e => e.Id) : 0L);
            return (lastId + increment);
        }

        #region Private members
        #endregion Private members
    }
}
